File: client-android/lib/src/main/kotlin/stasis/client_android/lib/model/server/devices/Device.kt
package stasis.client_android.lib.model.server.devices

import com.squareup.moshi.Json
import com.squareup.moshi.JsonClass
import stasis.client_android.lib.model.core.NodeId
import stasis.client_android.lib.model.server.users.UserId
import java.math.BigInteger
import java.time.Duration
import java.time.Instant
import java.util.UUID

@JsonClass(generateAdapter = true)
data class Device(
    val id: DeviceId,
    val name: String,
    val node: NodeId,
    val owner: UserId,
    val active: Boolean,
    val limits: Limits?,
    val created: Instant,
    val updated: Instant
) {
    @JsonClass(generateAdapter = true)
    data class Limits(
        @field:Json(name = "max_crates")
        val maxCrates: Long,
        @field:Json(name = "max_storage")
        val maxStorage: BigInteger,
        @field:Json(name = "max_storage_per_crate")
        val maxStoragePerCrate: BigInteger,
        @field:Json(name = "max_retention")
        val maxRetention: Duration,
        @field:Json(name = "min_retention")
        val minRetention: Duration
    )
}

typealias DeviceId = UUID

File: client-ios/lib/Sources/StasisClientLib/Model/Server/Devices/Device.swift
import Foundation

public typealias DeviceId = UUID

public struct Device: Sendable, Equatable, Hashable, Codable {
    public let id: DeviceId
    public let name: String
    public let node: NodeId
    public let owner: UserId
    public let active: Bool
    public let limits: Limits?
    public let created: Date
    public let updated: Date

    public init(
        id: DeviceId,
        name: String,
        node: NodeId,
        owner: UserId,
        active: Bool,
        limits: Limits?,
        created: Date,
        updated: Date
    ) {
        self.id = id
        self.name = name
        self.node = node
        self.owner = owner
        self.active = active
        self.limits = limits
        self.created = created
        self.updated = updated
    }

    public struct Limits: Sendable, Equatable, Hashable, Codable {
        public let maxCrates: Int64
        public let maxStorage: Int64
        public let maxStoragePerCrate: Int64
        public let maxRetention: SecondsDuration
        public let minRetention: SecondsDuration

        public init(
            maxCrates: Int64,
            maxStorage: Int64,
            maxStoragePerCrate: Int64,
            maxRetention: SecondsDuration,
            minRetention: SecondsDuration
        ) {
            self.maxCrates = maxCrates
            self.maxStorage = maxStorage
            self.maxStoragePerCrate = maxStoragePerCrate
            self.maxRetention = maxRetention
            self.minRetention = minRetention
        }
    }
}
